refactor: wallet: return util::Result from `GetReservedDestination` methods

theStack opened the pull request:
This PR is a follow-up to #25218, as suggested in comment https://github.com/bitcoin/bitcoin/pull/25218#discussion_r907710067. The interfaces of the methods `ReserveDestination::GetReservedDestination`, `{Legacy,Descriptor,}ScriptPubKeyMan::GetReservedDestination` are improved by returning `util::Result<CTxDestination>` instead of `bool` in order to get rid of the two `CTxDestination&` and `bilingual_str&` out-parameters.

theStack committed:
refactor: wallet: return util::Result from `GetReservedDestination` methods

furszy commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -179,7 +179,7 @@ class ScriptPubKeyMan
     virtual bool Encrypt(const CKeyingMaterial& master_key, WalletBatch* batch) { return false; }
 
-    virtual bool GetReservedDestination(const OutputType type, bool internal, CTxDestination& address, int64_t& index, CKeyPool& keypool, bilingual_str& error) { return false; }
+    virtual BResult<CTxDestination> GetReservedDestination(const OutputType type, bool internal, int64_t& index, CKeyPool& keypool) { return {}; }
```
Because CTxDestination default value is `CNoDestination`, this will return a succeed `BResult` object.
Would be better to do:
```c+++
return Untranslated("Not supported");
```

(same as we are doing for the base `GetNewDestination`)

MarcoFalke replied:
I would have expected for `{}` to return an empty error message. Is there any way to avoid this confusion at compile time?

ryanofsky replied:
> I would have expected for `{}` to return an empty error message. Is there any way to avoid this confusion at compile time?

I think the silent conversion of bilingual_str into errors is an anti-feature. Not every translated string is necessarily an error string, and the conversion makes it not possible to use a `BResult<bilingual_str>` type that returns a normal translated string. #25665 improves this by adding an explicit error contructor

furszy replied:
> I would have expected for `{}` to return an empty error message. Is there any way to avoid this confusion at compile time?

Yeah @MarcoFalke, my bad here, I shouldn't had reviewed the PR before sleep...

As we internally use an `std::variant`, and the first element is the `bilingual_str`. The `{}` will indeed initialize the variant with an empty error message.

What I tried to say on the message is that I don't think that having empty error strings are helpful. The caller cannot know where/why the function failed (which is the main purpose to return an error string in the first place).

Thinking out loud without test it:
Probably we could prevent this behavior by removing the empty constructor or by asserting that an error string cannot be empty.

> I think the silent conversion of bilingual_str into errors is an anti-feature. Not every translated string is necessarily an error string, and the conversion makes it not possible to use a BResult<bilingual_str> type that returns a normal translated string. https://github.com/bitcoin/bitcoin/pull/25665 improves this by adding an explicit error contructor

You know, the initial version that did for the generic `CallResult` was actually implemented in that way. I was using an `OperationResult<T, E> ErrorOut(string)` helper function instead of the silent conversion.

I have 25665 on my review list 👍🏼 . The last time that I checked it (brief check, thus why I haven't commented yet), I wasn't so sure about the internal std:.variant removal because every result was going to have the succeed object initialized AND the errors AND warnings in memory. And conceptually, a function can succeed OR fail. Not both.

But.. will comment there soon :). Need to finish up few other reviews first. Been focused on the nice, and long, #24914 lately.

theStack replied:
Changed to return `return Untranslated("Not supported");` for the sake of being consistent with other methods in the same class (GetNewDestination) right now, as suggested by furszy. The discussion here about general improvements to BResults seem to better fit to #25665, this PR is merely a refactor.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -2423,27 +2419,28 @@ std::set<std::string> CWallet::ListAddrBookLabels(const std::string& purpose) co
-        if (!m_spk_man->GetReservedDestination(type, internal, address, nIndex, keypool, error)) {
-            return false;
+
+        auto op_address = m_spk_man->GetReservedDestination(type, internal, nIndex, keypool);
+        if (op_address.HasRes()) {
+            address = op_address.ReleaseObj();
+        } else {
+            return op_address.GetError();
         }
         fInternal = keypool.fInternal;
```
What about this:
```c++
const auto& op_address = m_spk_man->GetReservedDestination(type, internal, nIndex, keypool);
if (!op_address) return op_address;
fInternal = keypool.fInternal;
address = op_address.ReleaseObj();
```

theStack replied:
Thanks, done.

theStack commented on src/wallet/spend.cpp:
```diff
@@ -706,9 +706,11 @@ static BResult<CreatedTransactionResult> CreateTransactionInternal(
+        } else {
+            error = _("Transaction needs a change address, but we can't generate it.") + Untranslated(" ") + op_dest.GetError();
         }
         scriptChange = GetScriptForDestination(dest);
```
Thanks, done. Scrictly speaking moving the scriptChange assignment into the conditional is a bit more than a refactor (it could have been done also before the BResult change), but it also makes sense to not unnecessarily call `GetScriptForDestination` in the failure case.

furszy reviewed (approved):
Code review ACK 224e1e1f

theStack commented:
Rebased on master, `util::Result` is used instead of `BResult` now accordingly. Also squashed the two commits into one as I think it's easier to review.

furszy reviewed (approved):
ACK 76b3c37f

jonatack commented:
Post-merge ACK
